WIP: 3038267 price trait

safalliajoseph opened the pull request:
Hey @krystalcode - I have added the price trait for commerce attribute. I have altered the product attribute entity and related classes heavily in the module. I'm not sure whether it is the correct approach. Could you please review. Thanks!

safalliajoseph committed:
Implement hook_entity_type_build
Alter the commerce_product_attribute and commerce_product_attribute_value entity class
Define new class for the commerce_product_attribute and commerce_product_attribute_value entity which extends
CommerceBundleEntityBase

safalliajoseph committed:
Alters the Attributes events and attribute forms

Alter the commerce_product.attribute_field_manager to use the one provided
by commerce_attributes module.
Extends the attribute add form to specify the new attribute entity
Specify a new class for product attribute event handler

safalliajoseph committed:
Use the ProductAttributeInterface and remove the unwanted class overrides

safalliajoseph committed:
Add edit and add forms for Product Attribute Value

safalliajoseph committed:
Add a view which lists the attribute values for a particular attribute type

safalliajoseph commented:
Hey @krystalcode, There is one scenario which came to my mind. Lets say a product variation has two attributes - Size and color.
During size and color insert, the price of size and color attributes will be added to this product variation.

Lets say we are editing the size attribute, How should the product variation price be updated? 
should it be,
current product variation price + price of attribute which is getting updated?
Which is clearly not correct. 

What are your thoughts?

safalliajoseph commented:
got answer from Dimitris for the [above](https://github.com/krystalcode/drupal8-commerce-attributes/pull/1#issuecomment-472388000) concern. We will be setting  a base cost field in product and will be calculating the product variation price with the base product cost and attribute price.

safalliajoseph committed:
Implement hooks to listen to attribute updates
Fetch all variation associated with that attribute and update the price
Implement hook to listen to varaition update and update the variation price

krystalcode commented on commerce_attributes.info.yml:
```diff
@@ -0,0 +1,7 @@
+name: Commerce Attributes
+type: module
+description: Provides price, stock and weight fields for commerce product attributes.
```
**CHANGE REQUEST**

"Provides price, stock and weight management at the product attribute level."

safalliajoseph replied:
Done.

krystalcode commented on commerce_attributes.module:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * @file
+ * Commerce Attributes module.
```
**CHANGE REQUEST**

"Hooks and functionality for the Commerce Attribute module."

safalliajoseph replied:
Done

krystalcode commented on commerce_attributes.module:
```diff
@@ -0,0 +1,41 @@
+ */
+
+/**
+ * Implements hook_entity_type_build().
```
**CHANGE REQUEST**

"Implements hook_entity_type_alter()"

safalliajoseph replied:
Done

krystalcode commented on commerce_attributes.module:
```diff
@@ -0,0 +1,41 @@
+    $priceHandler = \Drupal::service('commerce_attributes.variation_price_handler');
+    $priceHandler->updateProductVariationsPrice($entity);
+  }
+}
```
**CHANGE REQUEST**

*Coding standards*

In Drupal all variables are named `snake_case` while all class properties are named `camelCase` i.e. `$attribute_price`, `$price_handler`.

*Application logic*

You need first to check if the attribute (bundle of attribute value) has the trait enabled. If not, nothing to do here.

*Application logic*

It is possible that an attribute value had a price and then it was removed e.g. we were charging extra for an attribute but later on we stopped charging. We still need to update the prices in those cases.

The most efficient way would be to detect if the attribute value price has changed. In Drupal 7 I believe the original entity was available in this hook; I'm not sure about Drupal 8. If not available, we unfortunately have to update in all cases.

safalliajoseph replied:
Modified the variable name to use snake_case
_Application logic_
> You need first to check if the attribute (bundle of attribute value) has the trait enabled. If not, nothing to do here.

@krystalcode I have added the check to see if price field exists in the attribute value. There is a bug in the Trait-Form, and the added traits are not getting saved to the entity. I spent some time and could not figure out the issue. I was thinking maybe we can discuss this in our next weekly.

> It is possible that an attribute value had a price and then it was removed e.g. we wer ..

Added a condition by fetching the original entity and checked if there is a price change. Product variation price update will happen only if there is a price change in the attribute

krystalcode commented on commerce_attributes.module:
```diff
@@ -0,0 +1,41 @@
+  $priceHandler = \Drupal::service('commerce_attributes.variation_price_handler');
+  $newPrice = $priceHandler->getProductVariationPrice($entity);
+  $entity->setPrice($newPrice);
+}
```
**CHANGE REQUEST**

*Readability*

* We know the entity will always be a product variation. We can simply name `$variation` instead of `$entity`.
* Follow Drupal standards for naming function variables i.e. `$price_handler`, `$new_price`. 

*Application logic*

As with the attribute above, we need to check whether it's applicable to update the price for the given variation. So, I'm thinking, the trait we have added to the Attribute entity type defines whether attribute values of this type add an extra cost to the variation's price. Attributes can then be assigned to multiple variation types. The most flexible solution is then to also define a trait at the variation type. When selected no fields will be added to the variation (we already have the price field), it just means that prices for variations of this type will be calculated by our module. Variations of types that do not have the trait enabled will be left untouched by our module.

So:

* Add an "Attribute-based pricing" trait for variation types.
* On this hook, check if the given variation is of a type that has the trait enabled. If not, nothing to do.
* On the variation form, disable editing the price field if the variation is of a type that has the trait enabled.

safalliajoseph replied:
_Readability_ Done

_Application logic_ I have added attribute price trait to variation type. Please review. I was thinking if it would be best to add the trait to product type instead of variation type.

krystalcode commented on commerce_attributes.routing.yml:
```diff
@@ -0,0 +1,14 @@
+    _entity_form: commerce_product_attribute_value.edit
+    _title: 'Edit Attribute Value'
+  requirements:
+     _permission: 'access commerce administration pages'
```
**CHANGE REQUEST**

There is the `administer commerce_product_attribute` for managing attributes and their values. Also, titles by convention only have the first word starting with a capital letter (look at core modules e.g. "Add content", not "Add Content").

Nevertheless, you shouldn't need to define the routing for an entity's add/edit forms. Look at other entities to see if you're missing anything in the entity's annotations. The entity's route provider creates the routes for an add page, add form and edit form.

safalliajoseph replied:
Modified the permission and titles.

The route was not accessible without adding them in routing.yml file. I checked other product_attribute entity and its using `route_provider` handler for generating the routes.

krystalcode commented on config/install/views.view.attribute_values.yml:
```diff
@@ -0,0 +1,423 @@
+        - url
+        - url.query_args
+      tags:
+        - 'config:field.storage.commerce_product_attribute_value.price'
```
**CHANGE REQUEST**

*Security*

The view should be restricted by permission `administer commerce_product_attribute`.

safalliajoseph replied:
Done. Modified the permission to be `administer commerce_product_attribute`

krystalcode commented on commerce_attributes.module:
```diff
@@ -0,0 +1,41 @@
+  $entity_types['commerce_product_attribute']->setFormClass('add', 'Drupal\commerce_attributes\Form\ProductAttributeForm');
+  $entity_types['commerce_product_attribute']->setFormClass('edit', 'Drupal\commerce_attributes\Form\ProductAttributeForm');
+  $entity_types['commerce_product_attribute_value']->setFormClass('add', 'Drupal\commerce_attributes\Form\ProductAttributeValueForm');
+  $entity_types['commerce_product_attribute_value']->setFormClass('edit', 'Drupal\commerce_attributes\Form\ProductAttributeValueForm');
```
**CHANGE REQUEST**

We are overriding the attribute/value entity classes and we define the form classes there. You shouldn't need to override the form classes here as well; have you tested removing all `->setFormClass` from here?

safalliajoseph replied:
Yes. Without setting it in the alter hook, its not taking up and system is throwing errors

krystalcode commented on src/Entity/ProductAttribute.php:
```diff
@@ -0,0 +1,121 @@
+ *     "id",
+ *     "label",
+ *     "elementType",
+ *     "traits"
```
**CHANGE REQUEST**

The `CommerceBundleEntityBase` class that we're now extending is providing a `locked` property that should be exported here as well.

safalliajoseph replied:
Added the `locked` property

krystalcode commented on src/Entity/ProductAttribute.php:
```diff
@@ -0,0 +1,121 @@
+   *
+   * @var string
+   */
+  protected $label;
```
**CHANGE REQUEST**

Remove properties already defined in the base class.

safalliajoseph replied:
Done

krystalcode commented on src/Entity/ProductAttribute.php:
```diff
@@ -0,0 +1,121 @@
+namespace Drupal\commerce_attributes\Entity;
+
+use Drupal\Core\Entity\EntityStorageInterface;
+use Drupal\commerce\Entity\CommerceBundleEntityBase;
```
**CHANGE REQUEST**

We have a convention to list `use` statements in the following order:

* Application specific modules.
* Contrib modules.
* Drupal core dependencies.
* External dependencies.

Within each of these groups, they should sorted alphabetically. If we have many (like, more than 5, separate each group using an empty line). 

That would make it:

```
use Drupal\commerce\Entity\CommerceBundleEntityBase;
use Drupal\Core\Entity\EntityStorageInterface;
```

safalliajoseph replied:
Made the modifications. I looked into the [d.o coding standard doc](https://www.drupal.org/docs/develop/coding-standards/namespaces) and could not find anything which says about any particular order for use statements. Is there any doc which I can refer to?

krystalcode replied:
@safalliajoseph This is a convention that we follow in our team, it's not a generic Drupal coding standards. It makes it more readable and easy to find the `use` statement you're looking in some files where there are a lot of statements.

krystalcode commented on src/Entity/ProductAttributeValue.php:
```diff
@@ -0,0 +1,167 @@
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Field\BaseFieldDefinition;
+use Drupal\commerce\Entity\CommerceContentEntityBase;
+use Drupal\commerce_product\Entity\ProductAttributeValueInterface;
```
**CHANGE REQUEST**

Reorder `use` statements as explained in the comment above.

safalliajoseph replied:
Done!

krystalcode commented on src/Entity/ProductAttributeValue.php:
```diff
@@ -0,0 +1,167 @@
+ *   },
+ *    links = {
+ *     "add-form" = "/admin/commerce/product-attributes/manage/{commerce_product_attribute}/values/add",
+ *     "edit-form" = "/admin/commerce/product-attributes/manage/{commerce_product_attribute}/values/{commerce_product_attribute_value}/edit",
```
**CHANGE REQUEST**

With the previous UI we had a custom Remove button on the Attribute edit form. With the new UI, don't we need the delete form to delete attribute values?

safalliajoseph replied:
Yeah, Yes. I have added the content entity delete form for product attribute value.

krystalcode commented on src/Entity/ProductAttributeValue.php:
```diff
@@ -0,0 +1,167 @@
+ *   field_ui_base_route = "entity.commerce_product_attribute.edit_form",
+ * )
+ */
+class ProductAttributeValue extends CommerceContentEntityBase implements ProductAttributeValueInterface {
```
**CHANGE REQUEST**

Settings the forms and their links seem to be the only changes you are making compared to the original class, right? We don't need to extend `CommerceContentEntityBase` for the purposes of this module, it adds some translation capabilities which are irrelevant to this module - we don't need to get into changing the default behavior in relation to that.

Have a look if you can set the forms and their links with hooks; that class will not be needed then.

safalliajoseph replied:
Yes you are right. removed the ProductAttributeValue class for commerce_attribute module.

safalliajoseph replied:
I added the product attribute value entity back, as for the delete to properly work I had to alter the `toUrl` method a bit

krystalcode commented on src/Form/ProductAttributeForm.php:
```diff
@@ -0,0 +1,454 @@
+   *
+   * @var \Drupal\commerce\EntityTraitManagerInterface
+   */
+  protected $traitManager;
```
**CHANGE REQUEST**

You don't need to redefine properties already defined in the base class.

safalliajoseph replied:
Done

krystalcode commented on src/Form/ProductAttributeForm.php:
```diff
@@ -0,0 +1,454 @@
+    // be created until the attribute is saved.
+    if (!$attribute->isNew()) {
+      $form = $this->buildValuesForm($form, $form_state);
+    }
```
**CHANGE REQUEST**

We now have a separate page for managing Values, I thought that we would be removing the values form here? In that case remove all related functions below as well.

safalliajoseph replied:
Done. Removed values from product attribute screeen

krystalcode commented on src/Form/ProductAttributeForm.php:
```diff
@@ -0,0 +1,454 @@
+  public function save(array $form, FormStateInterface $form_state) {
+
+    $status = $this->entity->save();
+    $this->submitTraitForm($form, $form_state);
```
**CHANGE REQUEST**

*Readability*

Functions should not start with an empty line.
Break related code into blocks with comments as needed i.e.

```
function() {
$status = $this->entity->save();

// Process traits.
$this->submitTraitForm($form, $form_state);

$original_variation_types = ...
}
```

safalliajoseph replied:
done

krystalcode commented on src/Form/ProductAttributeValueForm.php:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    return $entity;
+  }
```
Not sure why you needed to override the default `getEntityFromRouteMatch` function, did you have a specific problem?

safalliajoseph replied:
`ProductAttributeValueForm` is something which I added. 
commerce_product module does not gives any form for ProductAttributeValue entity

krystalcode commented on src/Plugin/Commerce/EntityTrait/CommerceAttributePrice.php:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @CommerceEntityTrait(
+ *   id = "commerce_attribute_price",
+ *   label = @Translation("Has Price"),
```
**CHANGE REQUEST**

Label: "Adds price to the product's base price".

safalliajoseph replied:
Isn't the label which appears in product attribute add screen?
Enabling the trait which just add a price field to attribute. So I think `Has Price` label makes more sense. 
We have also added a trait to variation. Trait in variation is actually what determines whether the attribute price needs to be used in variation price calculation or not.  
Do you really want to change the label to "Adds price to the product's base price"

krystalcode commented on src/Plugin/Commerce/EntityTrait/CommerceAttributePrice.php:
```diff
@@ -0,0 +1,30 @@
+    $fields = [];
+    $fields['price'] = BundleFieldDefinition::create('commerce_price')
+      ->setLabel(t('Price'))
+      ->setRequired(TRUE);
```
**CHANGE REQUEST**

Not required. It's totally fine if some values do not add an extra price to the variation.
Why are you not defining extras e.g. description and display options? Look at the product variation price field.

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelInterface
+   */
+  protected $Logger;
```
**CHANGE REQUEST**

Properties should be `camelCase` i.e. `$logger` instead of `$Logger`.

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+   *
+   * @var \Drupal\Core\Entity\EntityManagerInterface
+   */
+  protected $entityManager;
```
**CHANGE REQUEST**

You don't need the entity manager; you already have the entity type manager to get the storages from. The entity manager simply calls the entity type manager internally in these cases for backwards compatibility reasons.

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+   * @param Drupal\commerce_attributes\Entity\ProductAttributeValue $attributeValue
+   *   The ProductAttributeValue entity.
+   */
+  public function updateProductVariationsPrice(ProductAttributeValue $attributeValue) {
```
**CHANGE REQUEST**

This is the only method that is called from outside of the class. It should be the only `public` method and should be the first one defined after the constructor. All other methods should follow and should be `protected`.

The method should also be called `updatePrices`; no need to include `ProductVariations` in it, we're in the `VariationPriceHandler` so it's implied we're updating the variation prices.

safalliajoseph replied:
Made the changes

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+
+    // Get all variation type associated with the attribute
+    // which is being updated.
+    $variationTypes = $this->getAllVariationTypes($attribute);
```
**CHANGE REQUEST**

All variables should be `snake_case` e.g. `$attribute_value`, `$variation_types`.

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+      ->condition($attributefieldName, $attributeValue);
+    $variations = $query->execute();
+    return $variations;
+  }
```
**CHANGE REQUEST**

* You're not getting one variation here, you get multiple; update the method name accordingly.
* No `Drupal::` in classes please. Add an extra property in the class `$variationStorage` and get it in the constructor from the entity type manager. You will then be able to use it in other functions as well.
* Many unused variables here - you increase memory consumption for no reason at all.

```
protected function getVariations(array $types, $attribute, $attribute_value) {
    // The attribute field name of variation type will always be in the
    // format of `attrubute_attribute_machine_name`.
    $attribute_field_name = 'attribute_' . $attribute;

    // Get all variations associated with an particular attribute value.
    return $this->variationStorage
      ->condition('type', $types, 'IN')
      ->condition($attribute_field_name, $attribute_value)
      ->execute();
  }
```

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+    foreach ($variations as $variation_id) {
+      $product_variation = $this->entityManager->getStorage('commerce_product_variation')->load($variation_id);
+      $product_id = $product_variation->get('product_id')->getValue()['0']['target_id'];
+      $product = $this->entityManager->getStorage('commerce_product')->load($product_id);
```
**CHANGE REQUEST**

`$product = $product_variation->get('product_id')->entity;` or `$product = $product_variation->get('product_id')->first()->entity;`

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+      $currency = $product->get('field_price')->getValue()['0']['currency_code'];
+      $newPrice = new Price((string) $newPrice, $currency);
+      $product_variation->setPrice($newPrice);
+      $product_variation->save();
```
**CHANGE REQUEST**

First, rename the field to just `price`.
Price objects are able to do maths. The attribute price sum should be a Price object. Then you get the price like:

`$base_price = $product->get('price')->toPrice();` or if you get an error ``$base_price = $product->get('price')->first()->toPrice();`

Then add them using the Price API:
`$new_price = $base_price->add($attributes_price);`

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+        $attributeTotalPrice += $attributePrice;
+      }
+    }
+    return $attributeTotalPrice;
```
**CHANGE REQUEST**

See above, use the Price API to do maths and return a Price object.

safalliajoseph replied:
Done

krystalcode commented on src/VariationPriceHandler.php:
```diff
@@ -0,0 +1,211 @@
+    return $newPrice;
+  }
+
+}
```
**CHANGE REQUEST**

Let's rename this function to `recalculatePrice`. Also, this is the 2nd API method (for external use) so move it to the top with the other one before protected methods.

safalliajoseph replied:
Renamed to `recalculateVariationPrice`

krystalcode commented:
On top of the comments, could we also rename the module to `commerce_attribute` to be inline with the singular naming convention in the commerce ecosystem e.g. `commerce_product`, `commerce_order` etc.

safalliajoseph committed:
Remane the module to commerce_attribute
Add a trait to product variation called attribute-pricing
Do the price calculations only for product variations which have attribute
pricing enabled

safalliajoseph committed:
Disable price field for Variation form if attribute pricing is enabled
Add uninstall hook and delete all configuration during uninstall

safalliajoseph committed:
Rename methhod name from getProductVariationPrice to getProductVariationPrice

safalliajoseph commented:
> On top of the comments, could we also rename the module to `commerce_attribute` to be inline with the singular naming convention in the commerce ecosystem e.g. `commerce_product`, `commerce_order` etc.

Done. Renamed to commerce_attribute

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+  $new_price = $attribute_value->get('price')->first()->toPrice()->getNumber();
+  if ($new_price == $original_price) {
+    return;
+  }
```
**CHANGE REQUEST**

*Performance/Readability*

The `$attribute_value_original` is only used to the get the price before the update. It does not contribute to readability i.e. `$attribute_value_original` is not that much more readable than `$attribute_value->original`, and there is no readability issue here in the first place. So there's no reason to allocate an unnecessary variable here.

*API use*

Then, let's use the Price API that provides functions for comparing prices.

With some refactoring to avoid code repetition and respecting line length soft limit, the above changes would be:

```
$original_price = $this->getPrice($attribute_value->original);
$new_price = $this->getPrice($attribute_value);
if (!$new_price->compareTo($original_price)) {
  return;
}
```
and
```
protected function getPrice(ProductAttributeValueInterface $value) {
  return $value->get('price')->first()->toPrice();
}
```

safalliajoseph replied:
I have modified the code. As this is in a module file, I do not think we can add protected functions. Do we need to define getPrice in a service class?

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+
+  // Updates the variation price associated with the attribute value.
+  $price_handler = \Drupal::service('commerce_attribute.variation_price_handler');
+  $price_handler->updateProductVariationsPrice($attribute_value);
```
**CHANGE REQUEST**

You don't need to declare a variable you're not going to use. With the suggested method name change, that would just be:

`\Drupal::service('commerce_attribute.variation_price_handler')->updatePrices($attribute_value);`

safalliajoseph replied:
Done

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+ */
+function commerce_attribute_commerce_product_variation_presave($variation) {
+  $bundle_entity_type_id = $variation->getEntityType()->getBundleEntityType();
+  $variation_type = \Drupal::service('entity_type.manager')->getStorage($bundle_entity_type_id)->load($variation->bundle());
```
**CHANGE REQUEST**

Line length limits please:

```
$variation_type = \Drupal::service('entity_type.manager')
  ->getStorage($bundle_entity_type_id)
  ->load($variation->bundle());
```

safalliajoseph replied:
Done

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+  // Updates the variation price to be the sum of its attributes which have has
+  // price trait enabled.
+  $price_handler = \Drupal::service('commerce_attribute.variation_price_handler');
+  $new_price = $price_handler->recalculateVariationPrice($variation);
```
**CHANGE REQUEST**

No need for storing the price handler to a variable, see comment above.

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+ */
+function commerce_attribute_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (substr($form_id, 0, 26) == 'commerce_product_variation') {
+    $product = \Drupal::routeMatch()->getParameter('commerce_product');
```
**CHANGE REQUEST**

We have the product variation entity from the form object (see code above), get the product from there: `$product = $variation->getProduct();`.

safalliajoseph replied:
Done

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    $product_type = Drupal::service('entity_type.manager')->getStorage('commerce_product_type')->load($product->bundle());
+    $variation_type = \Drupal::service('entity_type.manager')->getStorage('commerce_product_variation_type')->load($product_type->getVariationTypeId());
```
**CHANGE REQUEST**

Let's use existing APIs:

```
$product_type = $product->getEntityType();
$variation_type = $variation->getEntityType();
```

safalliajoseph replied:
We cannot use the above methods I think, as we are trying to load the bundle entity. I have modified the code to use the entity class for loading the entity

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+    if (!$variation_type->hasTrait('commerce_attribute_variation_price')) {
+      return;
+    }
+    $form['price']['widget']['0']['#disabled'] = TRUE;
```
**CHANGE REQUEST**

*Documentation*

So let's describe what we're doing here, why are we disabling price editing. Both in the code comments here, but also update the field description so that store managers understand.

safalliajoseph replied:
done

krystalcode commented on commerce_attribute.module:
```diff
@@ -0,0 +1,122 @@
+      $form['price']['widget']['0']['#default_value'] = [
+        'number' => $product->get('price')->first()->toPrice()->getNumber(),
+        'currency_code' => $product->get('price')->first()->toPrice()->getCurrencyCode(),
+      ];
```
**CHANGE REQUEST**

*Performance/Readability*

Here you're repeating all these function calls twice to get the price, and you violate coding standards. This is a case where it would be justified to use a temporary variable:

```
$price = $product->get('price')->first()->toPrice();
$form['price']['widget']['0']['#default_value'] = [
  'number' => $price->getNumber(),
  'currency_code' => $price->getCurrencyCode(),
];
```

However, check the Price API, there is an easier way:

```
$form['price']['widget']['0']['#default_value'] = $product
  ->get('price')
  ->first()
  ->toPrice()
  ->toArray();
```

safalliajoseph replied:
Done

krystalcode commented on commerce_attribute.services.yml:
```diff
@@ -0,0 +1,4 @@
+services:
+  commerce_attribute.variation_price_handler:
+    class: Drupal\commerce_attribute\VariationPriceHandler
+    arguments: ['@logger.factory', '@entity.manager', '@entity_type.manager', '@commerce_product.attribute_field_manager', '@entity.query']
```
**CHANGE REQUEST**

*Coding standards*

No need to have those super long lines here either:

```
arguments:
  - '@logger.factory'
  - '@entity.manager'
  - '@entity_type.manager'
  - '@commerce_product.attribute_field_manager'
  - '@entity.query'
```

safalliajoseph replied:
Done

safalliajoseph committed:
Coding standard fixes

safalliajoseph committed:
Coding standard fixes

krystalcode commented:
@safalliajoseph I have renamed the repository name/path to match the module's name, please updated your`.git/config` file.

krystalcode committed:
Added composer.json file

safalliajoseph committed:
Modify the routing name to be compatible with Drupal standards

safalliajoseph committed:
Use the perm administer commerce_product_attribute in attribute values view

safalliajoseph committed:
Add the locked property to the product attribute entity config export annotation

safalliajoseph committed:
Merge branch '3038267-price-trait' of github.com:krystalcode/drupal8-commerce-attributes into 3038267-price-trait

safalliajoseph committed:
Remove the property which are defined in the base class

safalliajoseph committed:
Re-arrange the use statements according to the Drupal standards

safalliajoseph committed:
Remove the ProductAttributeValue entity extend

safalliajoseph committed:
Remove the traitmanager property as its already defined in the base class

safalliajoseph committed:
Coding standard fixes

safalliajoseph committed:
Coding standard fixes

safalliajoseph committed:
Add a delete form for product attribute value
